components/de_amp/browser/de_amp_throttle.cc

```diff
@@ -0,0 +1,116 @@
+
+#include <utility>
+
+#include "base/feature_list.h"
+#include "brave/components/de_amp/browser/de_amp_url_loader.h"
+#include "brave/components/de_amp/common/features.h"
+#include "brave/components/de_amp/common/pref_names.h"
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/navigation_entry.h"
+#include "content/public/browser/page_navigator.h"
+#include "content/public/browser/web_contents.h"
+#include "mojo/public/cpp/bindings/pending_receiver.h"
+#include "mojo/public/cpp/bindings/pending_remote.h"
+#include "net/base/net_errors.h"
+#include "services/network/public/mojom/url_loader.mojom.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+#include "ui/base/page_transition_types.h"
+#include "ui/base/window_open_disposition.h"
+
+namespace de_amp {
+
+// static
+std::unique_ptr<DeAmpThrottle> DeAmpThrottle::MaybeCreateThrottleFor(
+    scoped_refptr<base::SequencedTaskRunner> task_runner,
+    const network::ResourceRequest& request,
+    const content::WebContents::Getter& wc_getter) {
+  auto* contents = wc_getter.Run();
+
+  if (!contents)
+    return nullptr;
+
+  PrefService* prefs =
+      user_prefs::UserPrefs::Get(contents->GetBrowserContext());
+
+  if (!base::FeatureList::IsEnabled(de_amp::features::kBraveDeAMP) ||
+      !prefs->GetBoolean(kDeAmpPrefEnabled)) {
+    return nullptr;
+  }
+
+  return std::make_unique<DeAmpThrottle>(task_runner, request, wc_getter);
+}
+
+DeAmpThrottle::DeAmpThrottle(
+    scoped_refptr<base::SequencedTaskRunner> task_runner,
+    const network::ResourceRequest& request,
+    const content::WebContents::Getter& wc_getter)
+    : task_runner_(task_runner), request_(request), wc_getter_(wc_getter) {}
+
+DeAmpThrottle::~DeAmpThrottle() = default;
+
+void DeAmpThrottle::WillProcessResponse(
+    const GURL& response_url,
+    network::mojom::URLResponseHead* response_head,
+    bool* defer) {
+  VLOG(2) << "deamp throttling: " << response_url;
+  *defer = true;
+
+  mojo::PendingRemote<network::mojom::URLLoader> new_remote;
+  mojo::PendingReceiver<network::mojom::URLLoaderClient> new_receiver;
+  mojo::PendingRemote<network::mojom::URLLoader> source_loader;
+  mojo::PendingReceiver<network::mojom::URLLoaderClient> source_client_receiver;
+  DeAmpURLLoader* de_amp_loader;
+  std::tie(new_remote, new_receiver, de_amp_loader) =
+      DeAmpURLLoader::CreateLoader(weak_factory_.GetWeakPtr(), response_url,
+                                   task_runner_);
+  InterceptAndStartLoader(
+      std::move(source_loader), std::move(source_client_receiver),
+      std::move(new_remote), std::move(new_receiver), de_amp_loader);
+}
+
+void DeAmpThrottle::Redirect(const GURL& new_url) {
+  delegate_->CancelWithError(net::ERR_ABORTED);
+
+  auto* contents = wc_getter_.Run();
+
+  if (!contents)
+    return;
+
+  auto* entry = contents->GetController().GetVisibleEntry();
```

Comment:
Don't catch the idea of getting **visible** entry.. Is passing referrer works as expected?
I.e. click on `some.com` => `google.com/amp/another` =(de-amp)=> `another.com` with referrer some.com?

Reply:
the visible entry is the replacement for the active entry and the referrer typically gets blocked in `SanitizeForRequest` anyway because of the domain change. The visible entry is either the pending entry or the last committed entry. I think this is the safest method to use here.

Reply:
My question was why not `GetLastCommittedEntry`? In my case visible entry is `google.com/amp/another`(because of a pending navigation) but the referrer should be `some.com`

Reply:
but we don't want the referrer from the page we were on, we want the same referrer that `google.com/amp/another` would have had